docs/recipes/reducers/03-RefactoringReducersExample.md

```diff
@@ -0,0 +1,420 @@
+# Splitting Reducer Logic Using Functional Decomposition and Reducer Composition
+
+It may be helpful to see examples of what the different types of sub-reducer functions look like, how they fit together, and how a large single reducer function can be refactored into a composition of several smaller sub-reducers.  (**Note**: this example is deliberately written in a verbose style in order to illustrate the concepts and the process of refactoring, rather than perfectly concise code.)
+
+Let's say that our initial reducer looks like this:
+
+```js
+const initialState = {
+    visibilityFilter : 'SHOW_ALL',
+    todos : []
+};
+
+
+function appReducer(state = initialState, action) {
+    switch(action.type) {
+        case 'SET_VISIBILITY_FILTER' : { 
+            return Object.assign({}, state, {
+                visibilityFilter : action.filter
+            });
+        }
+        case 'ADD_TODO' : {
+            return Object.assign({}, state, {
+                todos : state.todos.concat({
+                    id: action.id,
+                    text: action.text,
+                    completed: false
+                })
+            });
+        }
+        case 'TOGGLE_TODO' : {
+            return Object.assign({}, state, {
+                todos : state.todos.map(todo => {
+                    if (todo.id !== action.id) {
+                      return todo;
+                    }
+
+                    return Object.assign({}, todo, {
+                        completed : !todo.completed
+                    })
+                  })
+            });
+        } 
+        case 'EDIT_TODO' : {
+            return Object.assign({}, state, {
+                todos : state.todos.map(todo => {
+                    if (todo.id !== action.id) {
+                      return todo;
+                    }
+
+                    return Object.assign({}, todo, {
+                        text : action.text
+                    })
+                  })
+            });
+        } 
+        default : return state;
+    }
+}
+
+```
+
+That function is fairly short, but already becoming pretty hard to read.  We're dealing with two different areas of concern (filtering vs our list of todos), the nesting is making the update logic harder to read, and it's not exactly clear what's going on everywhere.
+
+A good first step might be to break out a utility function to return a new object with updated fields:
+
+```js
+function updateObject(oldObject, newValues) {
+    // Copy fields from oldObject to a new empty object.
+    // Then copy fields from newValues to the new object, overwriting any existing fields.
+    // This gives us a new updated copy, without modifying the original
```

Comment:
The utility function is great, and should definitely stay.  I was referring to the comment.